tictactoe.py

```diff
@@ -88,28 +90,67 @@ def get_state(self, config=None):
             if all([c == player for c in config[:: 4]]) or \
                     all([c == player for c in config[2:7: 2]]):
                 return '%s wins' % player
+        # Moved to after the win check
+        # Reason: If the last move decides the victor, it should not be declared as Draw
+        if all([c != ' ' for c in config]):
+            return 'Draw'
         return 'Game not finished'
 
+    def perform_unbeatable_move(self, player):
+        rank = -INFINITY
+        unbeatable_config = self.config
+        for c_index, c in enumerate(self.config):
+            if c == ' ':
+                new_config = self.config[:c_index] + player + self.config[c_index + 1:]
+                new_rank = self.get_rank(new_config, player)
+                if new_rank > rank:
+                    rank = new_rank
+                    unbeatable_config = new_config
+        return unbeatable_config
+
+    def get_rank(self, config, player):
+        turn = self.whose_turn(config)
+        opponent = 'O' if player == 'X' else 'X'
+        result = self.get_state(config)
+
+        if ('%s wins' % player) == result:
+            return 1
+        elif ('%s wins' % opponent) == result:
+            return -1
+        elif 'Draw' == result:
+            return 0
+        else:
+            # Define -infinity for
+            rank = -INFINITY if turn == player else INFINITY
+            for c_index, c in enumerate(config):
+                if c == ' ':
+                    if turn == player:
+                        # Maximize for turns made by player
+                        rank = max(rank, self.get_rank(config[:c_index] + turn + config[c_index + 1:], player))
+                    else:
+                        # Minimize for turns made by opponent
+                        rank = min(rank, self.get_rank(config[:c_index] + turn + config[c_index + 1:], player))
+            return rank
+
 
 def play(player1='easy', player2='easy'):
     board = Board()
-    board.print()
+    board.print(board.config)
```

Comment:
passing an object variable to the same object's methods is a little absurd

Reply:
I was using the same method for printing config's that are generated within the min-max recursion (for debugging purposes). Will update the PR with its removal.

Reply:
Resolved here: bc6c50cd702c940c930ad708f5e11470987e4fb9